kore/src/Kore/Step/Simplification/Overloading.hs

```diff
@@ -15,125 +19,236 @@ import Control.Error
+        overloadingHead(inj'(overloadedChildren))
+
+  Then it reduces the initial problem to
+    overloadingTerm@(overloadingHead(overloadingChildren))
+   vs
+    overloadingHead(inj'(overloadedChildren))
+-}
+unifyOverloadingVsOverloaded
+    :: MonadSimplify unifier
+    => SimplifierVariable variable
+    => Symbol
     -> TermLike variable
-    -> Inj (TermLike variable)
-    -> MaybeT unifier (Pattern variable)
-overloadedAndEqualsOverloading
-    termMerger
-    first@(App_ firstHead _)
-    second@Inj { injChild }
-  | App_ secondHead secondChildren <- injChild
+    -> Application Symbol (TermLike variable)
+    -> Inj ()
+    -> UnifyOverloading unifier (TermLike variable)
+unifyOverloadingVsOverloaded
+    overloadingHead
+    overloadingTerm
+    (Application overloadedHead overloadedChildren)
+    injProto
   = do
-    OverloadSimplifier { isOverloading, resolveOverloading } <-
-        Simplifier.askOverloadSimplifier
-    if isOverloading firstHead secondHead
-    then equalInjectiveHeadsAndEquals
-            termMerger
-            first
-            (resolveOverloading injProto firstHead secondChildren)
-    else do
-        Monad.guard =<< isConstructorOrOverloaded secondHead
-        returnBottom "different injected ctor"
-  | Just typeName <- notUnifiableType injChild
-  = returnBottom typeName
-  where
-    injProto = () <$ second
-
-    returnBottom typeName =
-        Monad.Trans.lift $ do
-            explainBottom
-                ("Cannot unify overloaded constructor with " <> typeName <> ".")
-                first
-                (synthesize $ InjF second)
-            empty
-    notUnifiableType :: TermLike variable -> Maybe (Doc ())
-    notUnifiableType (DV_ _ _) = Just "injected domain value"
-    notUnifiableType (BuiltinBool_ _) = Just "injected builtin bool"
-    notUnifiableType (BuiltinInt_ _) = Just "injected builtin int"
-    notUnifiableType (BuiltinList_ _) = Just "injected builtin list"
-    notUnifiableType (BuiltinMap_ _) = Just "injected builtin map"
-    notUnifiableType (BuiltinSet_ _) = Just "injected builtin set"
-    notUnifiableType (BuiltinString_ _) = Just "injected builtin string"
-    notUnifiableType _ = Nothing
-overloadedAndEqualsOverloading _ _ _ = empty
+    OverloadSimplifier { isOverloaded, isOverloading, resolveOverloading }
+        <- Simplifier.askOverloadSimplifier
+    Monad.guard (isOverloaded overloadingHead)
+    isSecondHeadConstructor <- isConstructorOrOverloaded overloadedHead
+    Monad.guard isSecondHeadConstructor
+    let overloadedTerm' =
+            resolveOverloading injProto overloadingHead overloadedChildren
+    if isOverloading overloadingHead overloadedHead
+        then return $ Pair overloadingTerm overloadedTerm'
+        else throwBottom "different injected ctor"
+
+notUnifiableError
+    :: Monad unifier => TermLike variable -> UnifyOverloading unifier a
+notUnifiableError (DV_ _ _) = throwBottom "injected domain value"
```

Comment:
All these failures are because domain values and builtins cannot be overloaded, is that right?

Reply:
Yes.  But you should know better, as you've added this code ;-) fe47fbe340a395c020b1d7737e60167ad65c9e29